### PR TITLE
Add optional grid_weights to mask_and_reduce_metric

### PR DESCRIPTION
**Pull request by santhil-cyber**
## Describe your changes

Added an optional `grid_weights` parameter to [mask_and_reduce_metric](cci:1://file:///Users/santhil/neural-lam/neural_lam/metrics.py:20:0-52:28)
in [neural_lam/metrics.py](cci:7://file:///Users/santhil/neural-lam/neural_lam/metrics.py:0:0-0:0). When `average_grid=True` and `grid_weights`
is provided, the function computes a weighted mean over the grid
dimension instead of `torch.mean`. Weights are correctly subsetted when
a boolean [mask](cci:1://file:///Users/santhil/neural-lam/neural_lam/models/ar_model.py:206:4-211:54) is also applied.

**Motivation:** For global weather forecasting on regular lat-lon grids,
grid nodes represent different physical areas — a node at the equator
covers ~5× more area than one at 80°N. Without area weighting,
`torch.mean` over-represents polar regions. This is the standard
approach used by WeatherBench2 and operational NWP centres.

All 6 existing metric functions now accept and forward `grid_weights`
as a keyword argument. No existing function signatures are changed —
`grid_weights` defaults to `None`, preserving exact existing behavior.

Added [tests/test_metrics.py](cci:7://file:///Users/santhil/neural-lam/tests/test_metrics.py:0:0-0:0) with 14 tests covering backward
compatibility, weighted mean correctness, cos-latitude weighting,
mask+weights interaction, and all 6 metric functions.

No new dependencies required.

## Issue Link

Refs #313

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)

### CODE REVIEW

File: neural_lam/metrics.py
Comment (kshirajahere) [inline neural_lam/metrics.py:1]: Nice scoped PR @santhil-cyber , this looks clean and well-targeted.

One edge case I think is worth handling in `mask_and_reduce_metric(...)`:

- `grid_weights` may live on CPU while `metric_entry_vals` is on GPU during eval, which can trigger device mismatch at multiplication time.
- masked `grid_weights` could (in principle) sum to zero, causing a divide-by-zero in weighted mean.

Would it make sense to add a small guard/cast like:

```python
if grid_weights is not None:
    grid_weights = grid_weights.to(
        metric_entry_vals.device, dtype=metric_entry_vals.dtype
    )
    if mask is not None:
        grid_weights = grid_weights[mask]

if average_grid and grid_weights is not None:
    w = grid_weights.unsqueeze(-1)
    w_sum = w.sum()
    if w_sum <= 0:
        raise ValueError("grid_weights must have positive sum after masking")
    metric_entry_vals = (metric_entry_vals * w).sum(dim=-2) / w_sum 
```

Comment (santhil-cyber) [reply neural_lam/metrics.py:1]: Great catches @kshirajahere man!!— both are real edge cases I missed:

Device/dtype mismatch — you're absolutely right. During eval, grid_weights could easily be a CPU tensor while metric_entry_vals sits on GPU. The .to(device, dtype) cast should go in early, before any masking or multiplication.

Zero-sum after masking: unlikely in normal use, but a defensive ValueError is the right call. Silent inf from divide-by-zero would be much harder to debug downstream.

Your suggested guard looks correct to me: I'll add both the device cast and the w_sum <= 0 check in the next commit. Thanks for catching this before a reviewer has to!
